Documentation and Cleanup for Release (#1)
qir-alliance/pyqir: pyqir-jit/README.md

> # pyqir-jit

Reviewer: For consistency with the other readme:
```suggestion
# pyqir_jit
```

> 

Reviewer: ```suggestion

The 'pyqir_jit' package provides an easy way to execute generated QIR for the purpose of
1. easily testing and experimenting with QIR code
2. connecting it to low-level Python-based lab software such as e.g. [QCoDes.](https://qcodes.github.io/Qcodes/user/intro.html) 

It contains the necessary [just-in-time compilation](https://en.wikipedia.org/wiki/Just-in-time_compilation) infrastructure as well an extensibility mechanism to define what actions to perform when a gate is applied in Python.
```

## Examples

There are JIT examples in the repository:

> - [Bernstein-Vazirani](../examples/jit/bernstein_vazirani.py) ([Bernstein-Vazirani QIR](../examples/jit/bernstein_vazirani.ll))

Reviewer: both links currently give me a 404. I would also suggest to merge this with the sentence below.

Author: There is a mismatch of path resolution between local and on GH. This doc will be moving to a new file to cleanup support for pypi.

Reviewer: This will be done in a separate PR.

> Let's walk though the example. Suppose we have the generated IR (`bernstein_vazirani.ll`). We can evaluate it with the `NonadaptiveJit`, and `GateLogger` to print out a simple log of the quantum application.

Reviewer: E.g.
```suggestion
Let's look at how to log the gate sequence for the following example:
- [Bernstein-Vazirani](../examples/jit/bernstein_vazirani.py)
 We can evaluate the [generated QIR](../examples/jit/bernstein_vazirani.ll) with the `NonadaptiveJit`, and `GateLogger` to print out a simple log of the quantum application.
```

```python
from pyqir_jit import NonadaptiveJit, GateLogger

from pathlib import Path
import os

path = Path(__file__).parent
file = os.path.join(path, "bernstein_vazirani.ll")

jit = NonadaptiveJit()
logger = GateLogger()

print("# output from NonadaptiveJit returning the uninitialized output")
jit.eval(file, logger)

print("# output from GateLogger")
logger.print()
```

Would generate the output:

```
# output from NonadaptiveJit returning the uninitialized output
[[Zero, Zero, Zero, Zero, Zero, Zero, Zero, Zero]]
# output from GateLogger
qubits[9]
out[9]
x qubit[8]
h qubit[0]
h qubit[1]
h qubit[2]
h qubit[3]
h qubit[4]
h qubit[5]
h qubit[6]
h qubit[7]
h qubit[8]
cx qubit[2], qubit[8]
cx qubit[3], qubit[8]
h qubit[0]
h qubit[1]
h qubit[2]
h qubit[3]
h qubit[4]
h qubit[5]
h qubit[6]
h qubit[7]
measure qubits[0] -> out[0]
measure qubits[1] -> out[1]
measure qubits[2] -> out[2]
measure qubits[3] -> out[3]
measure qubits[4] -> out[4]
measure qubits[5] -> out[5]
measure qubits[6] -> out[6]
measure qubits[7] -> out[7]
measure qubits[8] -> out[8]
```

## Building and Testing

See [Building](../docs/building.md)

## Limitations

> - QIR entrypoint for JIT must be named `QuantumApplication__Run`

Reviewer: What happens if the entry point takes arguments?
What happens if there is a branching based on a measurement result?

Reviewer: I added a line that states that entry point arguments are not yet supported.

> - QIR must contain the defined runtime in [module.ll](../qirlib/src/module.ll)

Reviewer: I added a comment that this is included if the pyqir_generator package is used to generate the QIR.
